Composer/packages/client/src/utils/publishStatusPollingUpdater.ts

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import httpClient from './httpUtil';
+
+enum PollingStateEnum {
+  Stopped,
+  Running,
+}
+
+export enum ApiStatus {
+  Publishing = 202,
+  Success = 200,
+  Failed = 500,
+  Unknown = 404,
+  HeaderTooLarge = 431,
+  BadRequest = 400,
+}
+
+interface OnDataPayload {
+  botProjectId: string;
+  targetName: string;
+  apiResponse: any;
+}
+type OnDataHandler = (payload: OnDataPayload) => void;
+
+export class PublishStatusPollingUpdater {
+  private status: PollingStateEnum = PollingStateEnum.Stopped;
+  private timerId = 0;
+  private publishTargetName: string;
+  private botProjectId: string;
+  private pollingInterval = 10000;
+
+  constructor(botProjectId: string, publishTargetId: string) {
+    this.botProjectId = botProjectId;
+    this.publishTargetName = publishTargetId;
+  }
+
+  private async fetchPublishStatusData(botProjectId: string, publishTargetName: string, onData: OnDataHandler) {
+    if (typeof onData !== 'function') {
+      console.error(new Error('onData should be a function.'));
+      return;
+    }
+
+    // TODO: DONT set HTTP Status code with 500 404 directly!
+    const response = await httpClient
+      .get(`/publish/${this.botProjectId}/status/${this.publishTargetName}`)
+      .catch((reason) => reason.response);
+
+    onData({
+      botProjectId,
+      targetName: publishTargetName,
+      apiResponse: response,
+    });
+  }
+
+  async start(onData: OnDataHandler) {
+    if (!this.botProjectId || !this.publishTargetName) return;
+    if (this.status === PollingStateEnum.Running) return;
+
+    this.status = PollingStateEnum.Running;
+
+    this.fetchPublishStatusData(this.botProjectId, this.publishTargetName, onData);
+    this.timerId = window.setInterval(
+      async () => this.fetchPublishStatusData(this.botProjectId, this.publishTargetName, onData),
+      this.pollingInterval
+    );
+  }
+
+  stop() {
+    if (this.timerId) {
+      window.clearInterval(this.timerId);
+    }
+    this.timerId = 0;
+    this.status = PollingStateEnum.Stopped;
+  }
+
+  restart(onData: OnDataHandler) {
+    this.stop();
+    this.start(onData);
+  }
+
+  isSameUpdater(botProjectId, targetName) {
+    return this.botProjectId === botProjectId && this.publishTargetName === targetName;
+  }
+}
+
+export const pollingUpdaterList: PublishStatusPollingUpdater[] = [];
```

Comment:
what's this? and where is it used?

Reply:
This is the polling updater list for store updaters. Using global value insteads of component's value is because component will re-render when getting publish status, so the component's value like `ref` or `state` will reset too.

Reply:
I dont see it being used anywhere

Reply:
It is used in `publish.tsx`.
```
      // Start updaters
      botList
        .filter(
          (bot) => !!bot.publishTarget && !pollingUpdaterList.some((u) => u.isSameUpdater(bot.id, bot.publishTarget))
        )
        .forEach((bot) => {
          if (pollingUpdaterList.some((updater) => updater.isSameUpdater(bot.id, bot.publishTarget))) return;
          const updater = new PublishStatusPollingUpdater(bot.id, bot.publishTarget);
          pollingUpdaterList.push(updater);
          updater.start(onReceiveUpdaterPayload);
        });
```
and
```
    // Add new updater
    if (!pollingUpdaterList.some((u) => u.isSameUpdater(currentBotStatus.id, publishTarget))) {
      const newUpdater = new PublishStatusPollingUpdater(currentBotStatus.id, publishTarget);
      newUpdater.start(onReceiveUpdaterPayload);
      pollingUpdaterList.push(newUpdater);
    }
```